### PR TITLE
feat: pre call diagnostics 

### PR DESCRIPTION
**Pull request by eswarclynn**
### Details(context, link the issue, how was the bug fixed, what does the new feature do)

-
-

### CODE REVIEW

File: packages/hms-video-store/src/device-manager/DeviceManager.ts [hunk 1]
```diff
@@ -34,38 +34,40 @@ export class DeviceManager implements HMSDeviceManager {
...
-      }
-    });
+  constructor(private store: Store, private eventBus: EventBus, sendAnalytics = true) {
+    if (sendAnalytics) {
+      const isLocalTrackEnabled = ({ enabled, track }: { enabled: boolean; track: HMSLocalTrack }) =>
+        enabled && track.source === 'regular';
+      this.eventBus.localVideoEnabled.waitFor(isLocalTrackEnabled).then(async () => {
+        await this.enumerateDevices();
+        if (this.videoInputChanged) {
+          this.eventBus.deviceChange.publish({ devices: this.getDevices() } as HMSDeviceChangeEvent);
```
Comment (raviteja83) [inline packages/hms-video-store/src/device-manager/DeviceManager.ts:44]: I think this is used other places also without analytics. if this is not related to analytics, better to rename the flag,

File: packages/hms-video-store/src/diagnostics/ConnectivityCheck.ts [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    if (this.state === ConnectivityState.MEDIA_PUBLISHED && this.errors.length === 0) {
+      this.state = ConnectivityState.COMPLETED;
+    }
+    this.completionCallback?.(this.buildReport());
+    this.sdk.leave();
+  }
+
+  private buildReport(): ConnectivityCheckResult {
+    const connectionQualityScore = this.networkScores.reduce((a, b) => a + b, 0) / this.networkScores.length;
+    // const currRTCStats = this.sdk.getWebrtcInternals()?.getCurrentStats();
```
Comment (raviteja83) [inline packages/hms-video-store/src/diagnostics/ConnectivityCheck.ts:152]: is this not rquired? if not remove the commented code.

File: packages/hms-video-store/src/diagnostics/ConnectivityCheck.ts [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+  private networkScores: number[] = [];
+  private errors: HMSException[] = [];
+  private isAudioTrackCaptured = false;
+  private isVideoTrackCaptured = false;
+
+  private cleanupTimer?: number;
+  private timestamp = Date.now();
+
+  constructor(private sdk: HMSSdk) {}
+  onRoomUpdate(): void {}
```
Comment (raviteja83) [inline packages/hms-video-store/src/diagnostics/ConnectivityCheck.ts:29]: these are for future implmentation? all of them are blank?

File: packages/hms-video-store/src/diagnostics/HMSDiagnostics.ts [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+  async startCameraCheck(inputDevice?: string) {
+    if (!this.localPeer) {
+      throw new Error('Local peer not found');
+    }
+
+    const localTrackManager = new LocalTrackManager(
+      this.store,
+      {
+        onFailure: exception => {
+          throw exception;
```
Comment (raviteja83) [inline packages/hms-video-store/src/diagnostics/HMSDiagnostics.ts:50]: this might become unhandled exception. we can send in onError no?

File: packages/hms-video-store/src/diagnostics/HMSDiagnostics.ts [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    this.localPeer.role = {
+      ...baseRole,
+      publishParams: { ...baseRole.publishParams, allowed: ['video'] },
+    };
+    this.deviceManager.init();
+    const settings = new HMSTrackSettingsBuilder()
+      .video(new HMSVideoTrackSettingsBuilder().deviceId(inputDevice || 'default').build())
+      .build();
+    const tracks = await localTrackManager.getLocalTracks({ audio: false, video: true }, settings);
+    const track = tracks.find(track => track.type === 'video') as HMSLocalVideoTrack;
```
Comment (raviteja83) [inline packages/hms-video-store/src/diagnostics/HMSDiagnostics.ts:66]: track.getVideoTracks()[0]

File: packages/hms-video-store/src/diagnostics/constants.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
+import { HMSRole } from '../internal';
+
+export const baseRole: HMSRole = {
```
Comment (raviteja83) [inline packages/hms-video-store/src/diagnostics/constants.ts:3]: diagnosticsRole makes more sense

File: packages/hms-video-store/src/diagnostics/HMSDiagnostics.ts [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+    }
+    const response = await fetchWithRetry(
+      tokenAPIURL.toString(),
+      { method: 'GET' },
+      [429, 500, 501, 502, 503, 504, 505, 506, 507, 508, 509, 510, 511],
+    );
+
+    const data = await response.json();
+
+    if (!response.ok) {
```
Comment (raviteja83) [inline packages/hms-video-store/src/diagnostics/HMSDiagnostics.ts:150]: why not check this before the above line?

File: packages/hms-video-store/src/diagnostics/HMSDiagnostics.ts [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+      throw new Error('SDK not found');
+    }
+    const connectivityCheck = new ConnectivityCheck(this.sdk);
+    this.sdk.setConnectivityListener(connectivityCheck);
+
+    const authToken = await this.getDiagnosticsAuthToken(region);
+    this.sdk.join({ authToken, userName: 'diagonistic-test' }, connectivityCheck);
+  }
+
+  private async getDiagnosticsAuthToken(region?: string): Promise<string> {
```
Comment (raviteja83) [inline packages/hms-video-store/src/diagnostics/HMSDiagnostics.ts:137]: it can be just AuthToken as we are already in diagnostics

File: packages/hms-video-store/src/diagnostics/HMSDiagnostics.ts (hunk 7 above)
Comment (eswarclynn) [reply packages/hms-video-store/src/diagnostics/HMSDiagnostics.ts:150]: The error requires data.code and data.message

File: packages/hms-video-store/src/diagnostics/HMSDiagnostics.ts [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+    if (!this.localPeer) {
+      throw new Error('Local peer not found');
+    }
+    if (!track) {
+      throw new Error('No audio track found');
+    }
+
+    this.localPeer.audioTrack = track;
+
+    const mediaRecorder = new MediaRecorder(track.stream.nativeStream);
```
[the remark this answers is not in the document]
Comment (eswarclynn) [reply packages/hms-video-store/src/diagnostics/HMSDiagnostics.ts:96]: Seems widely available https://developer.mozilla.org/en-US/docs/Web/API/MediaRecorder#browser_compatibility

File: packages/hms-video-store/src/diagnostics/ConnectivityCheck.ts (hunk 3 above)
Comment (eswarclynn) [reply packages/hms-video-store/src/diagnostics/ConnectivityCheck.ts:29]: These are not required but non-nullable as per the interface

File: packages/hms-video-store/src/device-manager/DeviceManager.ts (hunk 1 above)
Comment (eswarclynn) [reply packages/hms-video-store/src/device-manager/DeviceManager.ts:44]: Think at least as per the usage in the DeviceManager constructor it's just for analytics

Comment (raviteja83) [reply packages/hms-video-store/src/device-manager/DeviceManager.ts:44]: this is listened in audioSinkManager as well. for analytics it should be analytics.publish no?

File: packages/hms-video-store/src/transport/index.ts [hunk 10]
```diff
@@ -549,6 +573,7 @@ export default class HMSTransport {
     for (const track of tracks) {
       try {
         await this.publishTrack(track);
+        this.connectivityListner?.onMediaPublished();
```
Comment (raviteja83) [inline packages/hms-video-store/src/transport/index.ts:576]: typo. Listener.

File: packages/hms-video-store/src/diagnostics/ConnectivityCheck.ts [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+import { HMSException, HMSTrack, HMSTrackType, HMSTrackUpdate, HMSUpdateListener } from '../internal';
+import { HMSSdk } from '../sdk';
+import { HMSPeer } from '../sdk/models/peer';
+
+export class ConnectivityCheck implements HMSConnectivityListener, HMSUpdateListener {
+  private wsConnected = false;
+  private websocketURL?: string;
+  private initConnected = false;
+
+  private isPublishICEConnected = false;
```
Comment (raviteja83) [inline packages/hms-video-store/src/diagnostics/ConnectivityCheck.ts:12]: we can store all this boolean values in initialState object and reuse when clearing. similar to how it is done in sdk.

File: packages/hms-video-store/src/diagnostics/ConnectivityCheck.ts [hunk 12]
```diff
@@ -0,0 +1,190 @@
+import { ConnectivityCheckResult, ConnectivityState, HMSConnectivityListener } from './interfaces';
+import { RTCIceCandidatePair } from '../connection/IConnectionObserver';
+import { HMSException, HMSTrack, HMSTrackType, HMSTrackUpdate, HMSUpdateListener } from '../internal';
+import { HMSSdk } from '../sdk';
+import { HMSPeer } from '../sdk/models/peer';
+
+export class ConnectivityCheck implements HMSConnectivityListener, HMSUpdateListener {
```
Comment (raviteja83) [inline packages/hms-video-store/src/diagnostics/ConnectivityCheck.ts:7]: should we create a new interface instead of adding all black methods?

File: packages/hms-video-store/src/diagnostics/HMSDiagnostics.ts [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+    }
+    const connectivityCheck = new ConnectivityCheck(this.sdk, progress, completed);
+    this.sdk.setConnectivityListener(connectivityCheck);
+
+    const authToken = await this.getAuthToken(region);
+    this.sdk.join({ authToken, userName: 'diagonistic-test' }, connectivityCheck);
+  }
+
+  /** @internal */
+  initLocalPeer() {}
```
Comment (raviteja83) [inline packages/hms-video-store/src/diagnostics/HMSDiagnostics.ts:126]: not needed?

File: packages/hms-video-store/src/diagnostics/interfaces.ts [hunk 14]
```diff
@@ -0,0 +1,86 @@
+import { RTCIceCandidatePair } from '../connection/IConnectionObserver';
+import { HMSException, HMSLocalAudioTrack, HMSLocalVideoTrack } from '../internal';
+
+export interface DeviceCheckReturn {
```
Comment (raviteja83) [inline packages/hms-video-store/src/diagnostics/interfaces.ts:4]: Result?

File: packages/hms-video-store/src/diagnostics/DiagnosticsStatsCollector.ts [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+    });
+  }
+
+  // eslint-disable-next-line complexity
+  buildReport(): DiagnosticsRTCStatsReport {
+    const publishRoundTripTime = getLastElement(this.peerStatsList)?.publish?.totalRoundTripTime || 0;
+    const subscribeRoundTripTime = getLastElement(this.peerStatsList)?.subscribe?.totalRoundTripTime || 0;
+    const roundTripTime = (publishRoundTripTime + subscribeRoundTripTime) / 2;
+
+    const audioPacketsReceived = getLastElement(this.remoteAudioTrackStatsList)?.packetsReceived || 0;
```
Comment (raviteja83) [inline packages/hms-video-store/src/diagnostics/DiagnosticsStatsCollector.ts:50]: better to take a second argument as a key and handle || 0 in the getLastElement as well.

File: packages/hms-video-store/src/diagnostics/HMSDiagnostics.ts [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+  ) {
+    if (!this.sdk) {
+      throw new Error('SDK not found');
+    }
+
+    const authToken = await this.getAuthToken(region);
+    const { roomId } = decodeJWT(authToken);
+
+    this.sdk?.store.setRoom(new HMSRoom(roomId));
+    await this.sdk.leave();
```
Comment (KaustubhKumar05) [inline packages/hms-video-store/src/diagnostics/HMSDiagnostics.ts:125]: Why do we call leave before joining?

File: packages/hms-video-store/src/diagnostics/interfaces.ts [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+
+export interface SignallingReport {
+  isConnected: boolean; // true if websocket was connected successfully
+  isInitConnected: boolean; // True if init call was successful
+  websocketUrl?: string; // websocket url
+}
+
+export interface MediaServerReport {
+  stats?: DiagnosticsRTCStatsReport; // represents the overall stats of the call
+  isPublishICEConnected: boolean; // True is ICE connected successfully for both publish and subscribe
```
Comment (KaustubhKumar05) [inline packages/hms-video-store/src/diagnostics/interfaces.ts:63]:
```suggestion
  isPublishICEConnected: boolean; // True if ICE connected successfully for both publish and subscribe
```

File: packages/hms-video-store/src/diagnostics/interfaces.ts [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+  isSubscribeICEConnected: boolean;
+  connectionQualityScore?: number; // Averga of all the downlink scores for this call for this peer
```
Comment (KaustubhKumar05) [inline packages/hms-video-store/src/diagnostics/interfaces.ts:65]:
```suggestion
  connectionQualityScore?: number; // Average of all the downlink scores for this call for this peer
```

File: packages/hms-video-store/src/transport/index.ts [hunk 19]
```diff
@@ -78,6 +79,7 @@ export default class HMSTransport {
   private publishStatsAnalytics?: PublishStatsAnalytics;
   private subscribeStatsAnalytics?: SubscribeStatsAnalytics;
   private maxSubscribeBitrate = 0;
+  private connectivityListner?: HMSConnectivityListener;
```
Comment (KaustubhKumar05) [inline packages/hms-video-store/src/transport/index.ts:82]:
```suggestion
  private connectivityListener?: HMSConnectivityListener;
```

File: packages/hms-video-store/src/transport/index.ts [hunk 20]
```diff
@@ -294,7 +297,7 @@ export default class HMSTransport {
               HMSConnectionRole.Publish,
               ErrorFactory.WebrtcErrors.ICEDisconnected(
                 HMSAction.PUBLISH,
-                `local candidate - ${this.publishConnection?.selectedCandidatePair?.local.candidate}; remote candidate - ${this.publishConnection?.selectedCandidatePair?.remote.candidate}`,
+                `local candidate - ${this.publishConnection?.selectedCandidatePair?.local?.candidate}; remote candidate - ${this.publishConnection?.selectedCandidatePair?.remote?.candidate}`,
```
Comment (KaustubhKumar05) [inline packages/hms-video-store/src/transport/index.ts:300]: Can be refactored into a func that returns this, has been used multiple times

File: packages/hms-video-store/src/diagnostics/ConnectivityCheck.ts [hunk 21]
```diff
@@ -0,0 +1,182 @@
...
+  onSessionStoreUpdate(): void {}
+  onPollsUpdate(): void {}
+  onWhiteboardUpdate(): void {}
+
+  private _state: ConnectivityState = ConnectivityState.STARTING;
+  private get state(): ConnectivityState {
+    return this._state;
+  }
+  private set state(value: ConnectivityState) {
+    if (value < this._state) {
```
[the remark this answers is not in the document]
Comment (eswarclynn) [reply packages/hms-video-store/src/diagnostics/ConnectivityCheck.ts:68]: There are race conditions in HMSTransport, we are making sure that we're always progressing forward

File: packages/hms-video-store/src/device-manager/DeviceManager.ts (hunk 1 above)
Comment (eswarclynn) [reply packages/hms-video-store/src/device-manager/DeviceManager.ts]: Removed the `sendAnalytics` condition

File: packages/hms-video-store/src/diagnostics/DiagnosticsStatsCollector.ts [hunk 22]
```diff
@@ -6,6 +6,9 @@ import { HMSSdk } from '../sdk';
 
 const isValidNumber = (num: number | undefined): boolean => !!num && !isNaN(num);
 const getLastElement = <T>(arr: T[]): T | undefined => arr[arr.length - 1];
+const calculateAverage = <T>(arr: T[], predicate: (val: T) => number | undefined): number =>
+  arr.reduce((acc, curr) => acc + (predicate(curr) || 0), 0) /
+  arr.filter(curr => isValidNumber(predicate(curr))).length;
```
Comment (raviteja83) [inline packages/hms-video-store/src/diagnostics/DiagnosticsStatsCollector.ts:11]: the filter can be store in a value and reuse for reduce and length.

File: packages/hms-video-store/src/reactive-store/HMSReactiveStore.ts [hunk 23]
```diff
@@ -133,6 +135,14 @@ export class HMSReactiveStore<T extends HMSGenericTypes = { sessionStore: Record
     return this.stats;
   };
 
+  getDiagnosticsSDK = (): HMSDiagnosticsInterface => {
```
Comment (raviteja83) [inline packages/hms-video-store/src/reactive-store/HMSReactiveStore.ts:138]: why not have this on actions?

Comment (eswarclynn) [reply packages/hms-video-store/src/reactive-store/HMSReactiveStore.ts:138]: Did not want to add any getters to actions due to the one-way data flow convention we're following

File: packages/hms-video-store/src/transport/index.ts [hunk 24]
```diff
@@ -549,6 +573,7 @@ export default class HMSTransport {
     for (const track of tracks) {
       try {
         await this.publishTrack(track);
+        this.connectivityListener?.onMediaPublished();
```
Comment (raviteja83) [inline packages/hms-video-store/src/transport/index.ts:576]: this should be after all the tracks no?

Comment (raviteja83) [reply packages/hms-video-store/src/transport/index.ts:576]: should we sned the type here? whether audio/video/screen?

**Review by raviteja83 — APPROVED** [review]
